### ISSUE TITLE
Dataset issues - non-symmetric intractions

### ISSUE THREAD

**Issue by linas** [MOZI-AI/annotation-scheme#119]
This is another dataset issue bug report.  Gene interactions should be symmetric. But not all of them are marked as such. The current dataset has 537839 interacting gene-pairs. Of these, approx 193653 do not have a symmetric partner.  There are two possible fixes for this: (1) use a SetLink not a ListLink (2) use a script to generate the missing ones.  I am using the following script to clean this up:
```
(define (symmetrize-gene-interactions)
"
  The gene interactions use the asymmetric ListLink to denote
  gene pairs. But gene interactions are symmetrix, so they should
  have used the SetLink. Oh well. At this time, it is convenient
  to use the ListLink during pattern searches; but in this case,
  the interactions should be symmetrized. That's what this does.
"
   (define interact-q
      (Bind
         (VariableList
            (TypedVariable (Variable "g1") (Type 'GeneNode))
            (TypedVariable (Variable "g2") (Type 'GeneNode)))
         (Present
            (Evaluation (Predicate "interacts_with")
               (List (Variable "g1") (Variable "g2"))))
         (Evaluation (Predicate "interacts_with")
            (List (Variable "g2") (Variable "g1")))))

   (define sym-set (cog-execute! interact-q))
   (format #t "Found ~A gene interactions\n"
      (length (cog-outgoing-set sym-set)))
   (cog-delete sym-set)
)
```

**Comment by tanksha** [comment]
@linas I will use SetLink as the it will help reduce the atomspace size and also will make the pattern matching query simpler than we have currently.

**Comment by linas** [comment]
@tanksha -- of all the bug reports, this is a low-priority one; some of the others are more important, I think I nominated a list of bugs by important in #98

**Comment by Habush** [comment]
@tanksha can you please explain your proposed solutions for the performance issue esp. for the `pathway-interactors` function under @linas's issue thread #98 so he can comment on it?

**Comment by tanksha** [comment]
@linas @Habush the `pathway-gene-interactors` function is one of the hangs and wanted to improve its performance by applying a PLN rule before doing the annotations which completes the graph with the links which will just be pulled later during the annotation. (Like the caching but not the same I guess)

The general functionality of the  `pathway-gene-interactors` function is for a pathway (in which the input gene is a member of)

- Finds two proteins under this pathway, then, finds the Genes they are expressed by and then check if those Genes have a biogrid interaction (if they are linked with `interacts_with` predicate)

if the above pattern is found, then returns the biogrid_interaction of the two genes and the pubmedID reference. 

This will be done for each of the pathways a given gene has.

The PLN rule that I wrote [here](https://github.com/tanksha/annotation-scheme/blob/master/annotation/pln_rule.scm#L267-L293) will do what the `pathway-gene-interactors` required to do, but it will be applied once before the annotation starts.

The rule will create a link between a gene and the biogrid_interaction_of_other_two_genes (where all three genes are under the same pathway) so that the  `pathway-gene-interactors` function will just search for such pattern. Check this (https://github.com/tanksha/annotation-scheme/blob/master/annotation/functions.scm#L615-L628)

I noticed two things though,
- It takes a while to apply the rule to the whole atomspace (I checked with the sample atomspace and it didnt take much time but with the whole atomspace, it has been more than an hour since I started it  and it didnt finish yet. FYI my system I am trying on is only 12GB RAM)

- As you suggested, the need for additional atom types (like for GO, Pathway, SMPDB) will help. As you see here https://github.com/tanksha/annotation-scheme/blob/master/annotation/pln_rule.scm#L294-L311,  am checking if they are a GO term or a pathway that a gene is a Member of.  

What do you think of the above approach?

**Comment by linas** [comment]
>  the pathway-gene-interactors function is one of the hangs

This was fixed two or three weeks ago.  There are patches for it. If you go through the list of bugs and patches, you will find it.

When I say that I was able to make the whole system run anywhere from 20x to 100x faster, I am not kidding. You have to go thourgh the rather long backlog of patches and issues and resolve them.

**Comment by linas** [comment]
> 12GB RAM

with all patches applied, the atomspace never gets larger than 5GB in size.

**Comment by linas** [comment]
> additional atom types (like for GO, Pathway, SMPDB)

Yes, I agree, this will help performance by avoiding senseless string searches and the associated overhead. However, this is a far more complex fix, than the currently open pull reqs and issues.  Most of the patches and suggestions are real easy and simple, and do not require large code changes.

Again - I recommend going through the existing issue backlog, in priority order (I gave a list of priorities in #98) and do those. They're simple, they're easy, they do not require a lot of work.  Once you get the simple things done, then you can move to the more complex cases, which require more difficult changes to the code.

**Comment by Habush** [comment]
> However, this is a far more complex fix, than the currently open pull reqs and issues

I thought adding new types is a simple thing to do, just specify it in an `atom_types.script` file and the cmake macros will do the rest like in the [agi-bio](https://github.com/opencog/agi-bio). So I don't understand how it is a complex fix? Can you please elaborate?

**Comment by linas** [comment]
well, its easy to add the types, but then you have to go and regenerate all of your datasets. (i.e. update the scripts that generate the datasets). Then you have to go into the annotation code, and rip out all the string compares .. and rip out all the `GroundedPredicateNode`s that do the string compares .. and fix all the patterns that use them, which means changing the arguments to all sorts of functions, and removing a bunch of other functions. And then you have to run tests to make sure everything still works. Its not intellectually difficult, its just a lot of work that might be error-prone.  But contrast, most of the open issues can be fixed in 10-15 minutes, or half-an-hour, plus whatever time it takes to run your unit tests.  So it's an order-of-magnitude difference in the amount of work needed.

**Comment by linas** [comment]
Anyway - there's also a meta-issue, involving performance tuning.   There are two ways to do performance tuning.  These are:

1. Study the code, think about it, and think about how to rearrange it to go faster.

2. Instrument the code, and identify those parts that use the greatest amount of CPU time. Focus on fixing those.

Approach 2 gives much much better results than approach 1.  One only has a finite amount of energy -- spend it on what matters.

**Comment by linas** [comment]
fixed in pull req #189
